plutus-chain-index-core/src/Marconi/Index/ScriptTx.hs

```diff
@@ -0,0 +1,194 @@
+{-# LANGUAGE DeriveGeneric              #-}
+{-# LANGUAGE DerivingStrategies         #-}
+{-# LANGUAGE GADTs                      #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE OverloadedStrings          #-}
+{-# LANGUAGE ScopedTypeVariables        #-}
+
+module Marconi.Index.ScriptTx where
+
+import Codec.Serialise (deserialiseOrFail)
+import Control.Lens.Operators ((^.))
+import Data.ByteString qualified as BS
+import Data.Foldable (forM_, toList)
+import Data.Maybe (catMaybes, fromJust)
+import Database.SQLite.Simple qualified as SQL
+import Database.SQLite.Simple.FromField qualified as SQL
+import Database.SQLite.Simple.ToField qualified as SQL
+import GHC.Generics (Generic)
+
+import Cardano.Api (SlotNo)
+import Cardano.Api qualified as C
+import Cardano.Api.Shelley qualified as Shelley
+import Cardano.Ledger.Alonzo.Language qualified as Alonzo
+import Cardano.Ledger.Alonzo.Scripts qualified as Alonzo
+import Cardano.Ledger.Core qualified
+import Cardano.Ledger.Crypto qualified as LedgerCrypto
+import Cardano.Ledger.Keys qualified as LedgerShelley
+import Cardano.Ledger.Shelley.Scripts qualified as LedgerShelley
+import Cardano.Ledger.ShelleyMA.Timelocks qualified as Timelock
+
+import RewindableIndex.Index.VSqlite (SqliteIndex)
+import RewindableIndex.Index.VSqlite qualified as Ix
+
+
+newtype Depth = Depth Int
+
+newtype ScriptAddress = ScriptAddress Shelley.ScriptHash
+  deriving (Show, Eq)
+newtype TxCbor = TxCbor BS.ByteString
+  deriving (Show)
+  deriving newtype (SQL.ToField, SQL.FromField)
+
+-- * SQLite
+
+data ScriptTxRow = ScriptTxRow
+  { scriptAddress :: !ScriptAddress
+  , txCbor        :: !TxCbor
+  } deriving (Generic)
+
+instance SQL.ToField ScriptAddress where
+  toField (ScriptAddress hash)  = SQL.SQLBlob . Shelley.serialiseToRawBytes $ hash
+instance SQL.FromField ScriptAddress where
+  fromField f = SQL.fromField f >>=
+    either
+      (const cantDeserialise)
+      (\b -> maybe cantDeserialise (return . ScriptAddress) $ Shelley.deserialiseFromRawBytes Shelley.AsScriptHash b) . deserialiseOrFail
+    where
+      cantDeserialise = SQL.returnError SQL.ConversionFailed f "Cannot deserialise address."
+
+instance SQL.ToRow ScriptTxRow where
+  toRow o = [SQL.toField $ scriptAddress o, SQL.toField $ txCbor o]
+
+-- * Indexer
+
+type Query = ScriptAddress
+type Result = [TxCbor]
+
+data ScriptTxUpdate = ScriptTxUpdate
+  { txScripts :: [(TxCbor, [ScriptAddress])]
+  , slotNo    :: !SlotNo
+  } deriving (Show)
+
+type ScriptTxIndex = SqliteIndex ScriptTxUpdate () Query Result
+
+
+toUpdate :: forall era . C.IsCardanoEra era => [C.Tx era] -> SlotNo -> ScriptTxUpdate
+toUpdate txs = ScriptTxUpdate txScripts'
+  where
+    txScripts' = map (\tx -> (TxCbor $ C.serialiseToCBOR tx, getTxScripts tx)) txs
+
+getTxBodyScripts :: forall era . C.TxBody era -> [ScriptAddress]
```

Comment:
I think this may be useful to someone else. Should we not move this to something like `CardanoAPI`?

Reply:
Yes! For now, I'd say we keep it here until we decide how to structure cardano-api related code. For example, might make sense to create a decidated `cardano-api-extended` package which contains a bunch of these functions which can then be pushed upstream when stabilized and tested.

Reply:
Maybe this could be done in another PR?

But I guess yes, if the function would just return `Shelley.ScriptHash` (`ScriptAddress` is a newtype around it) then I guess Cardano.Api.Shelley could be a good location for it.

Reply:
A related thought: if we will get a hoogle, then writing this function could be done in two steps: one would be `getTxBodyScriptHashes :: forall era . C.TxBody era -> [ScriptHash]` and the other one would be the function we currently have, which just wraps `ScriptHash` into a newtype. This way hooglers can find the function, and if useful enough, can move them into a more suitable module.

Reply:
I think we can leave it for another PR.